### PR TITLE
Fix: per-system optimizer states for batched relaxation of independent systems

### PR DESCRIPTION
**Pull request by samblau**
### Problem

When `relax_mlff` optimizes multiple independent structures (e.g., drug molecules), all systems share a single LBFGS optimizer state over the concatenated position vector. The L-BFGS `s·y` dot products in the two-loop recursion mix curvature information from unrelated potential energy surfaces, corrupting the inverse Hessian approximation. This causes 1/25 convergence instead of 22/25.

See companion issue #94 for a minimal pure-NumPy reproduction (`examples/shared_lbfgs_bug.py`).

### Context

This was discovered by Sam Blau and an AI assistant (Claude, via [Axon](https://www.mirrorphysics.com) by Mirror Physics) while benchmarking kUPS for batched molecular geometry optimization. We think kUPS is a promising framework and hope this is useful.

### Fix

After the batched gradient computation (one forward+backward pass over all concatenated atoms), split gradients by system, pad to a common shape, and run `jax.vmap(optimizer.update)` over independent per-molecule optimizer states. The entire step — gradient + split + pad + vmapped LBFGS + unpad + scatter — compiles into a single JIT'd kernel.

### Results (25 drug molecules, UMA-s-1p2 OMol, H100)

| Method | Converged (fmax < 0.01 eV/Å) | Wall time | Per-step |
|--------|-------------------------------|-----------|----------|
| ASE LBFGS (PyTorch, sequential) | 22/25 | 287s | 105ms/mol |
| `relax_mlff` (shared LBFGS) | 1/25 | 145s | — |
| **This PR (vmapped per-molecule LBFGS)** | **22/25** | **19s** | **94ms/all 25** |

Step counts match ASE exactly (e.g., tafamidis: 19 both, dapagliflozin: 40 both, lenalidomide: 37 both). The 3 unconverged molecules are the same ones ASE can't converge in 200 steps.

### What's included

- `examples/shared_lbfgs_bug.py` — minimal pure-NumPy reproduction of the shared-LBFGS convergence bug (no kUPS or GPU needed)
- `examples/relax_mlff_vmapped.py` — standalone proof-of-concept implementing the fix using kUPS's model loading and optimizer primitives

### Design notes

This PR adds example scripts rather than modifying kUPS internals directly, since we're external contributors and wanted to demonstrate the approach without presuming how it should integrate into the Table/Lens/Propagator architecture. The key pattern is:

```python
# After batched gradient: grad shape (N_total, 3)
# Split by system using static gather indices (JIT-friendly)
padded_grad = grad_extended[gather_indices] * grad_mask  # (B, max_atoms, 3)

# Per-molecule vmapped LBFGS (independent states)
vmapped_update = jax.vmap(lambda g, s, p: optimizer.update(g, s, p))
updates, new_states = vmapped_update(padded_grad, per_mol_states, padded_pos)

# Scatter back to concatenated positions
positions = positions.at[scatter_indices].add(flat_updates)
```

A natural integration into kUPS could be a `per_system_optimizer: bool` flag in `RelaxParameters` that switches the `RelaxationPropagator` to use vmapped states. The vmapped LBFGS kernel compiles in <1s and runs in <1ms — the overhead is negligible.

### CODE REVIEW

**Comment by n-gao** [comment]
Hi @samblau , thanks a lot for your contribution and the issue. Great work! Though for the solution I think we would like to strive for a non-vmapped solution to keep in the spirit of kUPS and composability. This most likely means that when computing vector/matrix norms in optimizers, we need to do it on a per system basis explicitly (like we do in the rest of the codebase). Using vmap to solve this problem would be incompatible with the rest of the code. 

We will look into it! :)

**Comment by n-gao** [comment]
Thanks a lot! We fixed this in #109!
